pull/github.go

```diff
@@ -309,44 +310,170 @@ func (ghc *GitHubContext) Reviews() ([]*Review, error) {
+				edge := q.Repository.DirectCollaborators.Edges[i]
+				name := u.GetV3Login()
+
+				p, err := ParsePermission(edge.Permission)
+				if err != nil {
+					return nil, errors.Wrapf(err, "%s", name)
+				}
+				directPerms[name] = p
 			}
-			for _, u := range users {
-				collaborators[u.GetLogin()] = coalescePermission(u.GetPermissions())
+			if !q.Repository.DirectCollaborators.PageInfo.UpdateCursor(qvars, "directCursor") {
+				complete++
 			}
-			if res.NextPage == 0 {
+
+			for i, u := range q.Repository.AllCollaborators.Nodes {
+				edge := q.Repository.AllCollaborators.Edges[i]
+				name := u.GetV3Login()
+
+				p, err := ParsePermission(edge.Permission)
+				if err != nil {
+					return nil, errors.Wrapf(err, "%s", name)
+				}
+
+				collaborators = append(collaborators, &Collaborator{
+					Name:       name,
+					Permission: p,
+				})
+
+				for _, src := range edge.PermissionSources {
+					slug := src.Source.Team.Slug
+					if slug != "" {
+						teamMembership[name] = append(teamMembership[name], slug)
+					}
+				}
+			}
+			if !q.Repository.AllCollaborators.PageInfo.UpdateCursor(qvars, "allCursor") {
+				complete++
+			}
+
+			if complete == 2 {
 				break
 			}
-			opts.Page = res.NextPage
+		}
+
+		teamPerms, err := ghc.Teams()
+		if err != nil {
+			return nil, err
+		}
+
+		isPermissionViaRepo := func(c *Collaborator) bool {
+			if directPerms[c.Name] >= c.Permission {
+				return true
+			}
+			for _, team := range teamMembership[c.Name] {
+				if teamPerms[team] >= c.Permission {
+					return true
+				}
+			}
+			return false
+		}
+
+		for _, c := range collaborators {
+			c.PermissionViaRepo = isPermissionViaRepo(c)
 		}
 		ghc.collaborators = collaborators
 	}
 	return ghc.collaborators, nil
 }
 
-func coalescePermission(perms map[string]bool) string {
-	// standardize to new-style values used by GraphQL and other endpoints
-	switch {
-	case perms["admin"]:
-		return "admin"
-	case perms["push"] || perms["write"]:
-		return "write"
-	case perms["pull"] || perms["read"]:
-		return "read"
+func (ghc *GitHubContext) CollaboratorPermission(user string) (Permission, error) {
+	if ghc.permissions == nil {
+		ghc.permissions = make(map[string]Permission)
+	}
+	if p, ok := ghc.permissions[user]; ok {
+		return p, nil
+	}
+
+	var q struct {
+		Repository struct {
+			Collaborators struct {
+				Edges []struct {
+					Permission string
+				}
+			} `graphql:"collaborators(query: $user)"`
+		} `graphql:"repository(owner: $owner, name: $name)"`
+	}
+	qvars := map[string]interface{}{
+		"owner": githubv4.String(ghc.owner),
+		"name":  githubv4.String(ghc.repo),
+		"user":  githubv4.String(user),
+	}
+
+	if err := ghc.v4client.Query(ghc.ctx, &q, qvars); err != nil {
+		return PermissionNone, errors.Wrap(err, "failed to get collaborator permission")
+	}
+
+	var p Permission
+	if len(q.Repository.Collaborators.Edges) > 0 {
+		var err error
+		p, err = ParsePermission(q.Repository.Collaborators.Edges[0].Permission)
```

Comment:
What happens if we have more than one `edge`?

Reply:
I was assuming this wouldn't happen because there's a query parameter for the user name and users must be unique. I guess I should check to make sure it doesn't do substring matches or something like that.

Reply:
Good catch, the `query` parameter is a substring filter. Fixed in the latest commit to iterate through the results to make sure we find an exact match.